Add surface coupling design document (#408)
E3SM-Project/Omega: components/omega/doc/design/Coupling.md

coupling field arrays, unpacks driver attribute-vector data into typed
Omega arrays, applies imported fields to internal forcing state,
accumulates export fields over each coupling interval, and repacks them
for the driver. The design cleanly separates driver-specific operations
(handled in a Fortran/C++ bridge layer) from the C++ class, is agnostic
to the underlying driver architecture, and is extensible for future
fields (BGC, waves, land-ice).

## 2 Requirements

### 2.1 Requirement: Own coupling field storage

The class must own all coupling fields as `Array1DReal` arrays indexed
over ocean cells, consistent with other Omega field storage.

### 2.2 Requirement: Clean driver boundary

The driver interface layer (Fortran/C++ bridge) must handle all
driver-specific operations (index lookups, attribute vector layouts,
MCT/MOAB protocol). The C++ class receives and returns plain arrays of
doubles with no direct dependencies on MCT, MOAB, or other coupling
infrastructure.

### 2.3 Requirement: Field system integration

All coupling fields must be registerable with Omega's Field system for
diagnostic output.

### 2.4 Requirement: Per-coupling-interval operations

> The class must provide methods to import fields, apply them to forcing

Reviewer: @andrewdnolan what does 'apply them to the forcing state' mean?  Does this mean just load the received fluxes to the forcing class member variables?

Author: Yes, exactly. I will reword this sentence to me a bit clearer about what I meant.

Author: Updated in [c687732](https://github.com/E3SM-Project/Omega/pull/408/commits/c6877322bca318369ff9249bfad8737ad9373203) to be more clear.

state, accumulate export quantities each ocean timestep, export
accumulated fields, and reset accumulators.

> 

Reviewer 2: A similar comment - while this is focused on the interface to the coupler, we also need to flesh out how this interacts with the rest of the surface forcing.  In particular, where it sits in the run driver, how the fields are combined with other options like restoring. Conceptually, coupling should be identical to standalone forcing with just the source of the data coming from a coupler rather than a file (with the obvious exceptions that we also need to pass stuff back and the fields can be somewhat different). 

Also, what will be the configuration options?

Reviewer 3: The idea is that the forcing class (under Omega) is identical for both standalone and coupled. In coupled, it's populated by the coupling that andrew is building. In standalone, it reads from file. I am not sure if this answers your question though?

Author: @philipwjones I addressed these missing pieces in  [e4b2ae2](https://github.com/E3SM-Project/Omega/pull/408/commits/e4b2ae2eb4f3746076f55964d2db84a4407cd8db). Curious to hear if it that's enough detail?

### 2.5 Requirement: Driver architecture agnosticism

The class must support both MCT and MOAB layouts without conditional
compilation. The data layout (MCT column-major vs MOAB field-major) is
configured at runtime through the bridge layer.

### 2.6 Requirement: Coupling interval tracking

The class must track the coupling interval using Omega's `Alarm` system,
creating an internal periodic alarm at initialization to manage when
per-coupling-interval operations should occur.

### 2.7 Requirement: Coupler conversion layer

Private conversion methods must handle unit conversions and state
variable transformations between Omega's internal representation and
coupler expectations (e.g., conservative temperature ↔ potential

> temperature).

Reviewer 4: Does the coupler actually know about potential temperature or just in situ?

Reviewer 4: I think I would change `potential` to `in situ` here and below.

### 2.8 Requirement: Extensible design

Adding new coupling fields must require only localized changes: add
array members and update import/export logic. No restructuring of class
interfaces or method signatures should be needed.

## 3 Algorithmic Formulation

No algorithms are required.

## 4 Design

### 4.1 Data types and parameters

#### 4.1.1 Parameters

An enum class specifies the coupled driver layout:
```cpp
enum class CouplingLayout { MCT, MOAB };
```

#### 4.1.2 Class/structs/data types

`SurfaceCoupling` lives in `src/ocn/` alongside `AuxiliaryState` and
`OceanState`. It has no direct driver dependencies. The bridge layer in
`src/drivers/coupled/` is the only code that calls it.

```cpp
class SurfaceCoupling {
 public:
   // Import fields (x2o): coupler → ocean
   Array1DReal WindStressZonal;        ///< Foxx_taux  [N m⁻²]

>    Array1DReal WindStressMeridional;   ///< Foxx_tauy  [N m⁻²]

Reviewer: Do we receive a surface stress from the sea ice into this variable as well?  If so I'd prefer to rename this to suraceStressZonal/Meridional

   Array1DReal EvaporationFlux;        ///< Foxx_evap  [kg m⁻² s⁻¹]
   Array1DReal SeaIceFreshWaterFlux;   ///< Fioi_meltw [kg m⁻² s⁻¹]
   Array1DReal IcebergFreshWaterFlux;  ///< Fioi_bergw [kg m⁻² s⁻¹]
   Array1DReal SeaIceSalinityFlux;     ///< Fioi_salt  [kg m⁻² s⁻¹]
   Array1DReal RiverRunoffFlux;        ///< Foxx_rofl  [kg m⁻² s⁻¹]
   Array1DReal IceRunoffFlux;          ///< Foxx_rofi  [kg m⁻² s⁻¹]
   Array1DReal IceFraction;            ///< Si_ifrac   [-]
   Array1DReal SeaIcePressure;         ///< Si_bpress  [Pa]
   Array1DReal AtmosphericPressure;    ///< Sa_pslv    [Pa]

   // Export fields (o2x): ocean → coupler
   Array1DReal SurfaceTemperature;        ///< So_t    [°C]
   Array1DReal SurfaceSalinity;           ///< So_s    [g kg⁻¹]
   Array1DReal SurfaceVelocityZonal;      ///< So_u    [m s⁻¹]
   Array1DReal SurfaceVelocityMeridional; ///< So_v    [m s⁻¹]
   Array1DReal SSH;                       ///< So_ssh  [m]
   Array1DReal SSHGradientZonal;          ///< So_dhdx [m m⁻¹]
   Array1DReal SSHGradientMeridional;     ///< So_dhdy [m m⁻¹]
   Array1DReal SeaIceFormationHeat;       ///< Fioo_q      [W m⁻²]

>    Array1DReal FrazilIceMass;             ///< Fioo_frazil [kg m⁻² s⁻¹]

Reviewer 3: I suspect this will be updated to add 2 extra frazil coupling fields?

Author: Yes! But for the sake of the design doc, I'm just using mpas-o fields for simplicity. We will update in the actual implementation.

Reviewer 3: And we will want to rename `SeaIceFormationHeat` as well most likely once it is defined as sign definite (i.e. melt potential)

   Array1DReal FreshWaterHeatFlux;        ///< Faoo_h2otemp[W m⁻²]

   // Public methods
   int importFromCoupler(const Real *x2oData, int NFields, int NCells);
   int exportToCoupler(Real *o2xData, int NFields, int NCells);
   int applyImportedState(AuxiliaryState *AuxState);
   int accumulateExportState(const OceanState *State,
                             const Array3DReal &TracerArray,
                             int TimeLevel);
   void resetAccumulators();

 private:
   int NAccumSteps = 0;
   Alarm CouplingAlarm;
   std::map<std::string, int> ImportIdx;
   std::map<std::string, int> ExportIdx;
   CouplingLayout DataLayout;

   static Real potTempToConservTemp(Real PotTemp, Real Salinity, Real Pressure);
   static Real conservTempToPotTemp(Real ConsTemp, Real Salinity, Real Pressure);
};
```

### 4.2 Methods

#### 4.2.1 Creation

> The `create` method allocates import/export arrays sized to `NCellsAll`, stores

Reviewer 2: Any reason why this should be NCellsAll - really the coupler only needs NCellsOwned (without ghost cell and boundary info)

Author: Thanks for catching that.

the field index maps and `DataLayout`, creates the coupling `Alarm`, and calls
`registerFields`. After creation, `create` internally calls
`accumulateExportState` once to populate the initial export state.

```cpp
SurfaceCoupling *SurfaceCoupling::create(
   const std::string &Name, int OcnId, const HorzMesh *Mesh,
   const std::map<std::string, int> &ImportIdx,
   const std::map<std::string, int> &ExportIdx,
   const TimeInterval &CplInterval, CouplingLayout Layout);
```

#### 4.2.2 Retrieval

```cpp
SurfaceCoupling *SurfaceCoupling::getDefault();
SurfaceCoupling *SurfaceCoupling::get(const std::string &Name);
```

#### 4.2.3 Import and Export

`importFromCoupler` unpacks the driver array into typed member arrays using
the name→column-index map. The stride calculation is layout-dependent: MCT
uses `col*NCells + i`, MOAB uses `i*NFields + col`. Unit conversions (e.g.,

> in situ → conservative temperature, shortwave clamping) are applied inline.

Reviewer: what is shortwave clamping?

Author: `x2o_Foxx_swnet` should be bounded at zero, but I think in practice there can slight roundoff issues. 

The current mpas-ocean clips `x2o_Foxx_swnet` at zero. I presume we'll want to do the same:
> https://github.com/E3SM-Project/Omega/blob/59b6ca20fca813ec36703058a4ae3199c9eff8c9/components/mpas-ocean/driver/ocn_comp_mct.F#L2522

Reviewer 3: Can you clarify if the coupling layer does any vertical distribution (i.e. for shortwave penetration or river runoff)? I suspect it belongs within the ocean model but may be worth clarifying?

Reviewer 2: @alicebarthel Coupler has no knowledge of vertical meshes so does not handle any vertical distribution and it must happen in the ocean.

Reviewer 3: That makes sense, thanks for clarifying @philipwjones!

Author: Added an extra sentence to explain a little bit more detail about why we force short wave radiation to be positive.

`exportToCoupler` packs export arrays into the driver array. State fields
are divided by `NAccumSteps` (interval mean); flux fields are packed directly
(interval total).

`applyImportedState` copies imported fields into the appropriate locations in
`AuxiliaryState` for use by ocean physics.

`accumulateExportState` adds current-step values to running sums each ocean
timestep.

`resetAccumulators` zeroes all export arrays and sets `NAccumSteps = 0`.

#### 4.2.4 Conversion Methods

Temperature conversions between conservative (Omega internal) and in situ
(coupler expectation) are handled as private static methods. Simple

> constant-factor conversions are applied directly in import/export methods.

Reviewer 3: Does that include the hfactor and sfactor? I implemented these in the ocean model (change from flux to tracer tendency) which is more intuitive to me. Can you clarify which "constant-factor conversions" you are thinking of?

Author: @alicebarthel I've removed this sentence, because I don't think there any others than what you described.

> constant-factor conversions are applied directly in import/export methods.

Reviewer 4: Similar for salinity conversion from absolute to practical?

Author: @alicebarthel can confirm and provide more details, but I believe the plan is to pass absolute salinity.

Reviewer 3: We had talked about coupling absolute salinity but given that the sea ice model uses SSS to pull up brine salinity (from upward Darcy velocities) and may still use SSS for their freezing point (under discussion), I am now thinking we indeed need to convert to practical salinity as well to avoid affecting the sea ice thermodynamics.

#### 4.2.5 Destruction and removal

```cpp
void SurfaceCoupling::erase(const std::string &Name);
void SurfaceCoupling::clear();
```

### 4.3 Driver Interface Bridge

The bridge layer in `src/drivers/coupled/` provides `extern "C"` entry points
called from the Fortran driver (`ocn_comp_mct.F90` or `ocn_comp_moab.F90`).
The coupling interval and ocean timestep are independent: the driver calls

> run once per coupling interval; the ocean takes many timesteps within it.

Reviewer: sorry if this is a naive question, but do we often have a case with many ocean timesteps per coupling interval?  I'm wondering if I'm not understanding this line.

Reviewer 3: Afaik not often but the capability exists, right? 
Maybe it's just about rewording that last comment?
```suggestion
run once per coupling interval; the ocean can take several timesteps within it.
```

Reviewer 2: It has been the case in the past where the ocean communicated with the coupler less frequently (even once a day in the early days). It also allows for better asynchronous coupling. Think it is still a feature we want to keep even though it can sometimes be problematic from a physical standpoint (still conserves but creates lags and other issues in the application of forcing).

Author: Reworded this sentence slightly for clarity. But yes, the ocean can take multiple timesteps within the coupling interval. If I understand correctly, this would most often be used when the ocean-coupling timestep differs from the coupling timestep used by the other components. 

The example @jonbob has cited was POP would couple every 6-hours, while the rest of the coupled system might used a coupling interval of 30 minutes. Presumably this is flexibility we'd like to retain in Omega.

#### 4.3.1 Name-based Field Index Mapping

At startup, the Fortran bridge populates parallel arrays of Omega field names
and driver column indices, then passes them to `omega_ocn_init`. C++ builds
`std::map<std::string, int>` objects for import/export lookups. This approach
is driver-agnostic: MCT and MOAB differ only in how column indices are
obtained. Field names are passed as fixed-length null-terminated character
arrays (standard Fortran-C interop), parsed by stepping through the flat
array at stride `FieldNameLen`.

### 4.4 Interval Accumulation

Export state fields are accumulated each ocean timestep and divided by
`NAccumSteps` at export time (interval mean). Export flux fields are
accumulated and packed directly without dividing (interval total).

> `resetAccumulators` is called by `omega_ocn_export` after packing.

Reviewer 5: Are you thinking that each timestep we would call `accumulateExportState` from the timestepper?

Reviewer 5: I'm also wondering if it makes sense to have two different methods, `accumulateExportState` that handles the state fields and applies an average over the timesteps and `accumulateExportFlux` that integrates the flux over the timesteps. I'm also assuming that the flux fields are not a part of `OceanState` and so `accumulateExportFlux` would have different inputs than those given for `accumulateExportState` above.

Author: With regards to when/where this will be called, I was not envisioning from the timestepper. I was imagining it would be called from the coupled overload to `ocnRun`. 

The time step loop would look something like:
```cpp
// fetch default OceanState, TimeStepper, and SurfaceCoupler
OceanState *DefOceanState   = OceanState::getDefault();
TimeStepper *DefTimeStepper = TimeStepper::getDefault();
SurfaceCoupler *DefSurfaceCoupler = SurfaceCoupler::getDefault();

// get coupling alarm 
Alarm *CouplingAlarm = DefTimeStepper->getCouplingAlarm();

// maybe these two should be combined (or called from each other), since they should always be called together.
DefSurfaceCoupler->importFromCoupler(....);
DefSurfaceCoupler->applyImportedState(....)'

while (Err == 0 && !(CouplingAlarm->isRinging())) {
   
   DefTimeStepper->doStep(DefOceanState, SimTime);

   DefSurfaceCoupler->accumulateExportState(....);
   
}

// if coupler tells us, force write restate stream

DefSurfaceCoupler->exportToCoupler(....);
```

I think what I've scoped out above matches the logic MPAS-Ocean's `ocn_run_mct`, see here:
> https://github.com/E3SM-Project/E3SM/blob/d3a56abfd22e156a9e41d417dbdeedb6fb54d4cb/components/mpas-ocean/driver/ocn_comp_mct.F#L1173

Maybe adding this rough sketch of the run loop makes sent to add design document.

Author: With regards to separate `accumulateExportFlux` and `accumulateExportState`, where flux fields are integrated and state fields are averaged, that's initially what I envisioned as well. 

But after looking at [`ocn_export_mct`](https://github.com/E3SM-Project/E3SM/blob/d3a56abfd22e156a9e41d417dbdeedb6fb54d4cb/components/mpas-ocean/driver/ocn_comp_mct.F#L3123), it doesn't seem like the breakdown between flux (integrated) and state (averaged) fields quite holds.  Maybe I'm wrong about that, @jonbob could probably help clarify. 

Either way I think the proposed `accumulateExportState` probably needs to renamed, since it kind of conflicts with the `OceanState`. Maybe we should go with something like `accumulateExportFields` (if we go with one single function) or `accumulateExportStateFields` and `accumulateExportFluxFields` (if we go with separate functions).

Reviewer 5: Thanks for clarifying. To me, the design doc scope could include how coupling routines will be called but you can make the call based on how you see the work being organized.

Reviewer 5: It looks like this isn't yet resolved. This is my current understanding:

(1) The `accumulateExportState` and `exportToCoupler` as currently proposed will produce time-averaged fields. This covers most but not all variables.
(2) There is at least one field that is communicated without time averaging to the coupler, `ssh` https://github.com/E3SM-Project/E3SM/blob/d3a56abfd22e156a9e41d417dbdeedb6fb54d4cb/components/mpas-ocean/driver/ocn_comp_mct.F#L3354. Specify if this variable will be touched only by `exportToCoupler` or if we need a separate `export` method.
(3) There is at least one field that is accumulated without averaging, `accumulatedFrazilIceMass` https://github.com/E3SM-Project/E3SM/blob/d3a56abfd22e156a9e41d417dbdeedb6fb54d4cb/components/mpas-ocean/driver/ocn_comp_mct.F#L3407. Specify if the strategy is just to skip averaging in `exportToCoupler` or if we need a separate `export` method.
(4) `accumulateExportState` needs to be renamed for clarity because not all accumulated fields correspond to state variables.

Author: @cbegeman thanks for catching the naming issue. I've opted for `accumulateExportFields`, I think it makes the most sense. With that I've also changed `applyImportedState` --> `applyImportState` so that the tense in consistent with the other function names. 

`accumulateExportFields` just does the accumulation each ocean timestep; independent of whether the field is averaged or pass as sum back to the coupler. I think that this is a cleaner approach that having a separate accumulate/export method for the `So` and `Fxoo` fields, since the breakdown of what's averaged vs. accumulated vs. instantaneous does not neatly follow the breakdown between stare and flux fields. 

The text has been amended to make this more clear. Please let me know if what I have makes sense and is sufficient.

Reviewer 5: Sounds good. Thanks! Resolving

(omega-design-ocn-coupler)=
# Surface Coupling

## 1 Overview

> The `SurfaceCoupling` class manages all fields exchanged between Omega and

Reviewer 3: ```suggestion
The `SfcCoupling` class manages all fields exchanged between Omega and
```
?
we are moving all the forcing terms to be named `SfcForcing...` 
Just a suggestion.

per-coupling-interval operations should occur.

### 2.7 Requirement: Coupler conversion layer

Private conversion methods must handle unit conversions and state
variable transformations between Omega's internal representation and
coupler expectations (e.g., conservative temperature ↔ in situ
temperature).

### 2.8 Requirement: Extensible design

Adding new coupling fields must require only localized changes: add
array members and update import/export logic. No restructuring of class
interfaces or method signatures should be needed.

## 3 Algorithmic Formulation

No algorithms are required.

## 4 Design

### 4.1 Data types and parameters

#### 4.1.1 Parameters

An enum class specifies the coupled driver layout:
```cpp
enum class CouplingLayout { MCT, MOAB };
```

The initial implementation will not have any runtime configuration options.

#### 4.1.2 Class/structs/data types

`SurfaceCoupling` lives in `src/ocn/` alongside `AuxiliaryState` and
`OceanState`. It has no direct driver dependencies. The bridge layer in
`src/drivers/coupled/` is the only code that calls it.

```cpp
class SurfaceCoupling {
 public:
   // Import fields (x2o): coupler → ocean

>    Array1DReal SurfaceStressZonal;      ///< Foxx_taux  [N m⁻²]

Reviewer: These are moving to SfcStressXXXX right?

   Array1DReal SurfaceStressMeridional; ///< Foxx_tauy  [N m⁻²]
   Array1DReal SWHeatFlux;              ///< Foxx_swnet [W m⁻²]
   Array1DReal LatentHeatFlux;          ///< Foxx_lat   [W m⁻²]
   Array1DReal SensibleHeatFlux;        ///< Foxx_sen   [W m⁻²]
   Array1DReal LWHeatFluxUp;            ///< Foxx_lwup  [W m⁻²]
   Array1DReal LWHeatFluxDown;          ///< Faxa_lwdn  [W m⁻²]
   Array1DReal SeaIceHeatFlux;          ///< Fioi_melth [W m⁻²]
   Array1DReal IcebergHeatFlux;         ///< Fioi_bergh [W m⁻²]
   Array1DReal SnowFlux;                ///< Faxa_snow  [kg m⁻² s⁻¹]
   Array1DReal RainFlux;                ///< Faxa_rain  [kg m⁻² s⁻¹]
   Array1DReal EvaporationFlux;         ///< Foxx_evap  [kg m⁻² s⁻¹]
   Array1DReal SeaIceFreshWaterFlux;    ///< Fioi_meltw [kg m⁻² s⁻¹]
   Array1DReal IcebergFreshWaterFlux;   ///< Fioi_bergw [kg m⁻² s⁻¹]

>    Array1DReal SeaIceSalinityFlux;      ///< Fioi_salt  [kg m⁻² s⁻¹]

Reviewer 3: ```suggestion
   Array1DReal SeaIceSaltFlux;      ///< Fioi_salt  [kg m⁻² s⁻¹]
```
Just a note for your implementation

   Array1DReal RiverRunoffFlux;         ///< Foxx_rofl  [kg m⁻² s⁻¹]
   Array1DReal IceRunoffFlux;           ///< Foxx_rofi  [kg m⁻² s⁻¹]
   Array1DReal IceFraction;             ///< Si_ifrac   [-]
   Array1DReal SeaIcePressure;          ///< Si_bpress  [Pa]
   Array1DReal AtmosphericPressure;     ///< Sa_pslv    [Pa]

   // Export fields (o2x): ocean → coupler
   Array1DReal SurfaceTemperature;        ///< So_t    [°C]
   Array1DReal SurfaceSalinity;           ///< So_s    [g kg⁻¹]
   Array1DReal SurfaceVelocityZonal;      ///< So_u    [m s⁻¹]
   Array1DReal SurfaceVelocityMeridional; ///< So_v    [m s⁻¹]
   Array1DReal SSH;                       ///< So_ssh  [m]
   Array1DReal SSHGradientZonal;          ///< So_dhdx [m m⁻¹]
   Array1DReal SSHGradientMeridional;     ///< So_dhdy [m m⁻¹]
   Array1DReal SeaIceFormationHeat;       ///< Fioo_q      [W m⁻²]
   Array1DReal FrazilIceMass;             ///< Fioo_frazil [kg m⁻² s⁻¹]

>    Array1DReal FreshWaterHeatFlux;        ///< Faoo_h2otemp[W m⁻²]

Reviewer: I believe we also need to import WindSpeed10m from the coupler for KPP

                             int TimeLevel);
   void resetAccumulators();

 private:
   int NAccumSteps = 0;
   Alarm CouplingAlarm;
   std::map<std::string, int> ImportIdx;
   std::map<std::string, int> ExportIdx;
   CouplingLayout DataLayout;

   static Real potTempToConservTemp(Real PotTemp, Real Salinity, Real Pressure);
   static Real conservTempToPotTemp(Real ConsTemp, Real Salinity, Real Pressure);
};
```

### 4.2 Methods

#### 4.2.1 Creation

The `create` method allocates import/export arrays sized to `NCellsOwned`, stores
the field index maps and `DataLayout`, creates the coupling `Alarm`, and calls
`registerFields`. After creation, `create` internally calls
`accumulateExportState` once to populate the initial export state.

```cpp
SurfaceCoupling *SurfaceCoupling::create(
   const std::string &Name, int OcnId, const HorzMesh *Mesh,
   const std::map<std::string, int> &ImportIdx,
   const std::map<std::string, int> &ExportIdx,
   const TimeInterval &CplInterval, CouplingLayout Layout);
```

#### 4.2.2 Retrieval

```cpp
SurfaceCoupling *SurfaceCoupling::getDefault();
SurfaceCoupling *SurfaceCoupling::get(const std::string &Name);
```

> Other portions of the code can inquire whether running in coupled mode by

Reviewer: ```suggestion
Other portions of the code can inquire whether it is running in coupled mode by
```

doing:

```cpp
if (!SurfaceCoupling::getDefault) {
    // no instance of SurfaceCoupling, therefore running in standalone mode
}
```

#### 4.2.3 Import, Apply, Accumulate, Export

`importFromCoupler` unpacks the driver array into typed member arrays using
the name→column-index map. The stride calculation is layout-dependent: MCT
uses `col*NCells + i`, MOAB uses `i*NFields + col`. Unit conversions (e.g.,
in situ → conservative temperature) and corrections (e.g. forcing shortwave
radiation to be positive) are applied inline. The corrections are applied to

> address numerical issues, not and physical one, in the off chance monotonicity

Reviewer: I don't understand what you are saying here

step in agnostic of whether the fields are passed back to the coupler as
sums or averages.

`exportToCoupler` packs export arrays into the driver array. This function is
called directly after the coupled loop. State fields are divided by
`NAccumSteps` (interval mean); flux fields are packed directly (interval total);
and `SSH` is passed as an instantaneous field.

`resetAccumulators` zeroes all export arrays and sets `NAccumSteps = 0`.

A rough sketch of how/when these function will be called within `OcnRun` looks
like:

```cpp
// fetch default OceanState, TimeStepper, and SurfaceCoupler
OceanState *DefOceanState   = OceanState::getDefault();
TimeStepper *DefTimeStepper = TimeStepper::getDefault();
SurfaceCoupler *DefSurfaceCoupler = SurfaceCoupler::getDefault();

// get coupling alarm
Alarm *CouplingAlarm = DefTimeStepper->getCouplingAlarm();

// these two could be wrapped into a single call
DefSurfaceCoupler->importFromCoupler(...);
DefSurfaceCoupler->applyImportFields(...);

while (Err == 0 && !(CouplingAlarm->isRinging())) {

   DefTimeStepper->doStep(DefOceanState, SimTime);

   DefSurfaceCoupler->accumulateExportFields(...);

}

> // if coupler tells us, force write restate stream

Reviewer: ```suggestion
// if coupler tells us, force write restart stream
```
??
